subprojects/diagnostics/src/main/java/org/gradle/api/tasks/diagnostics/internal/configurations/model/ConfigurationReportModelFactory.java

```diff
@@ -0,0 +1,196 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.diagnostics.internal.configurations.model;
+
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.ConfigurationVariant;
+import org.gradle.api.artifacts.PublishArtifact;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.attributes.AttributeMatchingStrategy;
+import org.gradle.api.attributes.AttributesSchema;
+import org.gradle.api.capabilities.Capability;
+import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal;
+import org.gradle.api.internal.attributes.DefaultCompatibilityRuleChain;
+import org.gradle.api.internal.attributes.DefaultDisambiguationRuleChain;
+import org.gradle.api.internal.file.FileResolver;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * Factory for creating {@link ConfigurationReportModel} instances which represent the configurations present in a project.
+ */
+public final class ConfigurationReportModelFactory {
+    private final FileResolver fileResolver;
+
+    public ConfigurationReportModelFactory(FileResolver fileResolver) {
+        this.fileResolver = fileResolver;
+    }
+
+    public ConfigurationReportModel buildForProject(Project project) {
+        final ConfigurationRuleScanner scanner = new ConfigurationRuleScanner(project);
+        final List<ReportAttribute> attributesWithCompatibilityRules = scanner.getAttributesWithCompatibilityRules();
+        final List<ReportAttribute> attributesWithDisambiguationRules = scanner.getAttributesWithDisambiguationRules();
+
+        final Map<String, ReportConfiguration> convertedConfigurations = new ConcurrentHashMap<>(project.getConfigurations().size());
+        project.getConfigurations().stream()
+            .map(ConfigurationInternal.class::cast)
+            .forEach(configuration -> getOrConvert(configuration, project, convertedConfigurations));
+
+        return new ConfigurationReportModel(project.getName(),
+            convertedConfigurations.values().stream().sorted(Comparator.comparing(ReportConfiguration::getName)).collect(Collectors.toList()),
+            attributesWithCompatibilityRules, attributesWithDisambiguationRules);
+    }
+
+    private ReportConfiguration getOrConvert(ConfigurationInternal configuration, Project project, Map<String, ReportConfiguration> convertedConfigurations) {
+        return convertedConfigurations.computeIfAbsent(configuration.getName(), name -> {
```

Comment:
Even with a `Concurrent` hash map it isn't safe to recurse in `computeIfAbsent` -- you can deadlock. This code should:
- `get` it from the map
- if `null`, create and `put` it in
- return the non-null result
It's uglier but necessary to be safe.

Reply:
How can it deadlock though?  This map is a local variable only ever accessed by the current thread, so even if it acquires the lock on an extended configuration's key multiple times (if that configuration is reached via direct and transitive extensions when converting) this is fine - the same thread will (re)request the same lock and already hold it.

Now whether this actually breaks the contract of the method is another question.

The implementation of this method in `ConcurrentHashMap` notes that it throws an `IllegalStateException` if "the computation detectably attempts a recursive update to the map that would otherwise never complete" - that's not happening.  It also notes that the method "must not attempt to update any other mappings of this map" - which I'm not, I'm only _inserting_ new mappings which didn't previously exist, never modifying existing ones.  So for a *narrow* interpretation of "update", I'm adhering to the contract.

So my initial reading was that this was okay - `HashMap` aggressively checks for concurrent modifications and fails, but `ConcurrentHashMap`, which uses a more complex internal node structure, does not detect a failure.  Adding other keys on `HashMap` would definitely be wrong: https://bugs.openjdk.java.net/browse/JDK-8071667.  Doing the same on `ConcurrentHashMap` *may* be okay, *but may also be contingent on the specific keys being adding to the map in a specific order* and could suddenly break with more/different keys: https://stackoverflow.com/questions/28840047/recursive-concurrenthashmap-computeifabsent-call-never-terminates-bug-or-fea   Various SO questions point out the subtle differences on the javadoc between `Map`, `HashMap` and `ConcurrentHashMap` and seem to simultaneously saying this multiple recursive insert pattern is 1) allowed and functions fine using `CHM`, providing working example code or 2) is implicitly banned by the contract of the method and should always be avoided.

Not knowing if this currently functioning behavior is contingent on a certain type of inputs is a scary situation to be in, so I'm going to add an external static method per the workaround in that linked JDK bug which does more or less what you suggest and go back to using `HashMap`.

Thanks for flagging this!  I'd like to say I learned something here, but honestly I think I'm a little more confused than before I started looking into these method contracts.  Seems like if this type of thing is banned, it should be simple to add a fail-fast exception to the `computeIfAbsent` call in `CHM` which always fires and enforces the contract, even if a particular pattern of usage would be okay.  That such a check is missing I take as evidence that usage like mine isn't an _automatic_ violation of the contract...but at this point 🤷

Reply:
Not sure when exactly this happened, but by Java 17 [this Javadoc](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent(K,java.util.function.Function)) has been updated to indicate that _any_ modification of the underlying map is out-of-contract:
> The mapping function must not modify this map during computation.

So for future-proofing, this is against contract for sure.